File: project_pyratio/recipe_readers.py
# recipe_readers.py
# Code for parsing recipes from the recipe data file into Python types

from pathlib import Path
import re

from abbreviations import AbbreviationResolver
from databases import DataBase
from facilities import FacilityCategory
from products import ProductQuantity
from rates import Time
from rational_utilities import read_rational
from recipes import Recipe


############################################
# Recipe Block Specification for Data File #
############################################

# Any whitespace at the start or end of a line is stripped before parsing
# Recipe blocks are separated by one or more blank lines in the data file
# Any line beginning with a '#' is ignored (and not considered to be blank)

# A BNF-style description of the recipe block grammar:
#            recipe_block := outputs arrow_line inputs optional_nametag_line
#                 outputs := ingredient_line | ingredient_line outputs
#              arrow_line := "^" time_period "s" "(" facility_category ")"
#                  inputs := ingredient_line | ingredient_line inputs
#   optional_nametag_line := "" | "<>" recipe_name
#         ingredient_line := number product_name

COMMENT_LINE_REGEX: re.Pattern[str] = re.compile(r"^\#.*$")
ARROW_LINE_REGEX: re.Pattern[str] = re.compile(r"^\^ *([0-9.]+) *s *\((.+)\)$")
NAMETAG_LINE_REGEX: re.Pattern[str] = re.compile(r"^<> *(.+)$")
INGREDIENT_LINE_REGEX: re.Pattern[str] = re.compile(r"^([0-9]+) +(.+)$")


################################################
# Functions for Parsing the Entire Recipe File #
################################################

# This operates something like split(listoflines, isempty, keepempty = false)
# However, Base.split() only accepts sequences <: AbstractString as its input
# There really could exist a somewhat more general version of that function
# Why won't split just accept an arbitrary iterable and predicate function?
# And I can't seem to trick Julia into doing it by wrapping a Vector{String}
# It runs into all kinds of problems with AbstractChar and the SubString type
# This array-splitting problem also seems to be weirdly, totally un-Google-able
# Am I the only person who has ever wanted to split up an array like this???
def _split_on_blank_lines(list_of_lines: list[str]) -> list[list[str]]:
    list_of_blocks: list[list[str]] = []
    current_block_start: int | None = None

    for i, line in enumerate(list_of_lines):
        if line == "":
            if current_block_start is not None:
                block = list_of_lines[current_block_start:i]
                list_of_blocks.append(block)
            current_block_start = None
        elif current_block_start is None:
            current_block_start = i

    if current_block_start is not None:
        block = list_of_lines[current_block_start:]
        list_of_blocks.append(block)

    return list_of_blocks


def read_recipe_file(data_file_name: Path, db: DataBase) -> list[list[str]]:
    with open(data_file_name, "r", encoding="utf-8") as f:
        file_lines = map(str.strip, f.readlines())
    file_lines = list(filter(
        lambda line: re.match(COMMENT_LINE_REGEX, line) is None, file_lines
    ))

    all_products = set(p.name for p in db.products)
    product_resolver = AbbreviationResolver(all_products)

    all_categories = set(fc.name for fc in db.facility_categories)
    category_resolver = AbbreviationResolver(all_categories)

    for i, line in enumerate(file_lines):

        m = re.match(INGREDIENT_LINE_REGEX, line)
        if m is not None:
            number_string, product_name = m[1], m[2]
            if product_name not in all_products:
                full_product_name = product_resolver(productname)
                file_lines[i] = f"{number_string} {full_product_name}"
            continue

        m = re.match(ARROW_LINE_REGEX, line)
        if m is not None:
            seconds_string, category_name = m[1], m[2]
            if category_name not in all_categories:
                full_category_name = category_resolver(category_name)
                file_lines[i] = f"^ {seconds_string} s ({full_category_name})"

    return _split_on_blank_lines(file_lines)


#########################################
# Functions for Parsing a Single Recipe #
#########################################

def _parse_ingredient(
    recipe_block: list[str],
    i: int
) -> ProductQuantity[int] | None:
    if i < 0 or i >= len(recipe_block):
        return None

    m = re.match(INGREDIENT_LINE_REGEX, recipe_block[i])
    if m is None:
        return None

    number_string, product_name = m[1], m[2]
    return ProductQuantity(int(number_string), product_name)


def _parse_arrow_line(
    recipe_block: list[str],
    i: int
) -> tuple[Time, FacilityCategory]:
    if i < 0 or i >= len(recipe_block):
        raise ValueError("arrow line index out of bounds")

    m = re.match(ARROW_LINE_REGEX, recipe_block[i])
    if m is None:
        raise ValueError(f"unable to parse arrow line: {recipe_block[i]!r}")

    seconds_string, category_name = m[1], m[2]
    return Time(read_rational(seconds_string)), FacilityCategory(category_name)


def _parse_nametag_line(recipe_block: list[str], i: int) -> str | None:
    if i < 0 or i >= len(recipe_block):
        return None
    if i == len(recipe_block) - 1:
        raise ValueError("started to parse nametag line before end of recipe")

    m = re.match(NAMETAG_LINE_REGEX, recipe_block[i])
    if m is None:
        raise ValueError(f"unable to parse nametag line: {recipe_block[i]!r}")

    recipe_name = m[1]
    return str(recipe_name)


def read_recipe(recipe_block: list[str]) -> Recipe:
    i = 0

    r_outputs: list[ProductQuantity[int]] = []
    ingredient = _parse_ingredient(recipe_block, i)
    while ingredient is not None:
        r_outputs.append(ingredient)
        i += 1
        ingredient = _parse_ingredient(recipe_block, i)

    r_period, r_madein = _parse_arrow_line(recipe_block, i)
    i += 1

    r_inputs: list[ProductQuantity[int]] = []
    ingredient = _parse_ingredient(recipe_block, i)
    while ingredient is not None:
        r_inputs.append(ingredient)
        i += 1
        ingredient = _parse_ingredient(recipe_block, i)

    r_name = _parse_nametag_line(recipe_block, i)

    return Recipe(r_name, r_outputs, r_inputs, r_period, r_madein)

File: project_rustio/src/recipereaders.rs
// recipereaders.rs
// Code for parsing recipes from the recipe data file into Julia types

use std::collections::HashSet;
use std::fs;
use std::path::PathBuf;

use lazy_static::lazy_static;
use regex::Regex;

use crate::abbreviations::AbbreviationResolver;
use crate::databases::DataBase;
use crate::facilities::FacilityCategory;
use crate::products::{Product, ProductQuantity};
use crate::rates::Time;
use crate::rationalutilities::readrational;
use crate::recipes::Recipe;


//////////////////////////////////////////////
// Recipe Block Specification for Data File //
//////////////////////////////////////////////

// Any whitespace at the start or end of a line is stripped before parsing
// Recipe blocks are separated by one or more blank lines in the data file
// Any line beginning with a '#' is ignored (and not considered to be blank)

// A BNF-style description of the recipe block grammar:
//            recipe_block := outputs arrow_line inputs optional_nametag_line
//                 outputs := ingredient_line | ingredient_line outputs
//              arrow_line := "^" time_period "s" "(" facility_category ")"
//                  inputs := ingredient_line | ingredient_line inputs
//   optional_nametag_line := "" | "<>" recipe_name
//         ingredient_line := number product_name

lazy_static! {
	static ref COMMENTLINEREGEX: Regex = Regex::new(r"^\#.*$").unwrap();
	static ref ARROWLINEREGEX: Regex = Regex::new(r"^\^ *([0-9.]+) *s *\((.+)\)$").unwrap();
	static ref NAMETAGLINEREGEX: Regex = Regex::new(r"^<> *(.+)$").unwrap();
	static ref INGREDIENTLINEREGEX: Regex = Regex::new(r"^([0-9]+) +(.+)$").unwrap();
}

//////////////////////////////////////////////////
// Functions for Parsing the Entire Recipe File //
//////////////////////////////////////////////////

// This operates something like split(listoflines, isempty, keepempty = false)
// However, Base.split() only accepts sequences <: AbstractString as its input
// There really could exist a somewhat more general version of that function
// Why won't split just accept an arbitrary iterable and predicate function?
// And I can't seem to trick Julia into doing it by wrapping a Vector{String}
// It runs into all kinds of problems with AbstractChar and the SubString type
// This array-splitting problem also seems to be weirdly, totally un-Google-able
// Am I the only person who has ever wanted to split up an array like this???
fn splitonblanklines(listoflines: &Vec<String>) -> Vec<&[String]> {
	let mut listofblocks = Vec::new();
	let mut currentblockstart = None;
	
	for (i, line) in listoflines.iter().enumerate() {
		if line.is_empty() {
			if let Some(start) = currentblockstart {
				let block = &listoflines[start..i];
				listofblocks.push(block);
			}
			currentblockstart = None;
		} else if currentblockstart.is_none() {
			currentblockstart = Some(i);
		}
	}
	
	if let Some(start) = currentblockstart {
		let block = &listoflines[start..];
		listofblocks.push(block);
	}
	
	listofblocks
}

pub fn readrecipefile(datafilename: PathBuf, db: &DataBase) -> Vec<Vec<String>> {

	let mut filelines: Vec<String> = fs::read_to_string(datafilename)
	    .expect("should be able to open data file")
		.lines()
		.map(|line| line.trim())
		.filter(|&line| !COMMENTLINEREGEX.is_match(line))
		.map(|line| line.to_owned())
		.collect();
	
	let allproducts: HashSet<&str> = db.products
	    .iter().map(Product::name).collect();
	let mut productresolver = AbbreviationResolver::new(
		&allproducts
    );
	
	let allcategories: HashSet<&str> = db.facilitycategories
	    .iter().map(FacilityCategory::name).collect();
	let mut categoryresolver = AbbreviationResolver::new(
		&allcategories
	);
	
	for (i, line) in filelines.iter_mut().enumerate() {
		
		let m = INGREDIENTLINEREGEX.captures(line);
		if let Some(m) = m {
			let (_, [numberstring, productname]) = m.extract();
			if !allproducts.contains(productname) {
				let fullproductname = productresolver.call(productname);
				*line = format!("{} {}", numberstring, fullproductname);
			}
			continue;
		}

		let m = ARROWLINEREGEX.captures(line);
		if let Some(m) = m {
			let (_, [secondsstring, categoryname]) = m.extract();
			if !allcategories.contains(categoryname) {
				let fullcategoryname = categoryresolver.call(categoryname);
				*line = format!("^ {} s ({})", secondsstring, fullcategoryname);
			}
			continue;
		}
	}
	
	let recipeblocks = splitonblanklines(&filelines);
	recipeblocks.into_iter().map(Vec::from).collect()
}


///////////////////////////////////////////
// Functions for Parsing a Single Recipe //
///////////////////////////////////////////

fn parseingredient(recipeblock: &Vec<String>, i: usize) -> Option<ProductQuantity<i64>> {
	let line = match recipeblock.get(i) {
		Some(line) => line,
		None => return None,
	};

	match INGREDIENTLINEREGEX.captures(line) {
		Some(m) => {
			let (_, [numberstring, productname]) = m.extract();
			let quantity = numberstring.parse().unwrap();
			Some(ProductQuantity::with_productname(quantity, productname))
		},
		None => None,
	}
}

fn parsearrowline(recipeblock: &Vec<String>, i: usize) -> (Time, FacilityCategory) {
	let line = match recipeblock.get(i) {
		Some(line) => line,
		None => panic!("arrow line index out of bounds"),
	};

	match ARROWLINEREGEX.captures(line) {
		Some(m) => {
			let (_, [secondsstring, categoryname]) = m.extract();
			let t = Time::new(readrational(secondsstring));
			let fc = FacilityCategory::new(categoryname);
			(t, fc)
		},
		None => panic!("unable to parse arrow line: {:?}", line),
	}
}

fn parsenametagline(recipeblock: &Vec<String>, i: usize) -> Option<&str> {
	let line = match recipeblock.get(i) {
		Some(line) => line,
		None => return None,
	};

	if i + 1 != recipeblock.len() {
		panic!("started to parse nametag line before end of recipe");
	}

	match NAMETAGLINEREGEX.captures(line) {
		Some(m) => {
			let (_, [recipename]) = m.extract();
			Some(recipename)
		},
		None => panic!("unable to parse nametag line: {:?}", line),
	}
}

pub fn readrecipe(recipeblock: &Vec<String>) -> Recipe {
	let mut i = 0;
	
	let mut r_outputs = Vec::new();
	let mut ingredient = parseingredient(recipeblock, i);
	while let Some(quantity) = ingredient {
		r_outputs.push(quantity);
		i += 1;
		ingredient = parseingredient(recipeblock, i);
	}
	
	let (r_period, r_madein) = parsearrowline(recipeblock, i);
	i += 1;
	
	let mut r_inputs = Vec::new();
	let mut ingredient = parseingredient(recipeblock, i);
	while let Some(quantity) = ingredient {
		r_inputs.push(quantity);
		i += 1;
		ingredient = parseingredient(recipeblock, i);
	}
	
	let r_name = parsenametagline(recipeblock, i);
	
	Recipe::new(r_name.as_deref(), r_outputs, r_inputs, r_period, r_madein)
}
